statmoments/benchmark/benchlib.py

```diff
@@ -49,34 +49,35 @@ def create_engine(self, tr_len, cl_len):
 def benchmark(benchset):
   number, repeat = 1, 1
 
-# Name        Implementation                     MB  tr_cnt   tr_len  cl_cnt   tr/sec   time res_time
-# bivar_m1    bivar_vtk(m(1, 1))                 50    5000      500       1      725    6.9      0.0
-# bivar_m1    bivar_vtk(m(1, 1))                124    5000     1000       1      132   37.7      0.0
-  print('{:12}{:30}{:>7}{:>8}{:>9}{:>8}{:>8}{:>7}{:>9}'.format(
-        'Kernel', 'Implementation', 'MB', 'tr_cnt', 'tr_len', 'cl_cnt', 'tr/sec', 'time', 'res_time'))
+  print('{:12}{:30}{:>7}{:>8}{:>9}{:>8}{:>10}{:>11}{:>10}{:>7}'.format(
+    'Kernel', 'Implementation', 'MB', 'tr_cnt', 'tr_len', 'cl_cnt', 'tr/sec', 'avg_upd_ms', 'avg_tt_ms', 'time_s'))
 
   for name, params in benchset:
-    for fctr, tr_count, tr_len, cl_count in product(*params):
-      tproc, ttest = [], []
+    for engine_factory, tr_count, tr_len, cl_count in product(*params):
+      update_times, ttest_times = [], []
+      assert repeat > 0
       for _ in range(repeat):
-        # Garbage collect memory after the prev iteration to free the existing memory
-        traces, engine = None, None
-        traces, classifiers = fctr.create_data(tr_count, tr_len, cl_count)
-        engine = fctr.create_engine(tr_len, cl_count)
+        traces, classifiers = engine_factory.create_data(tr_count, tr_len, cl_count)
+        engine = engine_factory.create_engine(tr_len, cl_count)
         for _ in range(number):
-          start = timer()
+          start = perf_counter()
           engine.update(traces, classifiers)  # Streaming: layout and accumulator
-          tproc.append(timer() - start)
+          update_times.append(perf_counter() - start)
 
-          start = timer()
-          for _ in ttests(engine):           # On-demand: t-tests
-            _ = None
-          ttest.append(timer() - start)
+          start = perf_counter()
+          for _ in ttests(engine):  # On-demand: t-tests
+            pass
+          ttest_times.append(perf_counter() - start)
 
-      mproc, mtt = min(tproc), min(ttest)
-      max_mom = ''.join(map(str, [engine.moment]*2))
+      total_time = sum(update_times) + sum(ttest_times)
```

Comment:
`min` is taken to find out time with less interruptions. We need one fastest pass.
`total_time` doen't have much sense as `update` happens often and `ttest` rarely on demand.

Reply:
I find min unfair, it takes "min" and then assumes all of them were that fast.
Can look at actual distribution, maybe.

Reply:
min gives the performance upper bound